### PR TITLE
various memory optimizations

### PR DESCRIPTION
**Pull request by imatiach-msft**
various memory optimizations in interpret-community done as part of integration with another product
- where possible use _local_importance_values since local_importance_values converts from numpy to list
- avoid typed_dataset in some paths since that creates a new pandas dataframe
- add clear_references parameter to DatasetWrapper such that it can be cleared after use, which should reduce memory usage - note this is only for users who know what they are doing
- added test for dense and wide data which can be used for future performance testing as well

### CODE REVIEW

File: python/interpret_community/dataset/dataset_wrapper.py [hunk 1]
```diff
@@ -312,9 +317,12 @@ def string_index(self, columns=None):
...
+        # Temporarily convert to pandas for easier and uniform string handling,
+        # only use top sampled rows to limit memory usage for string type test
         if isinstance(self._dataset, np.ndarray):
-            tmp_dataset = pd.DataFrame(self._dataset, dtype=self._dataset.dtype)
+            tmp_dataset = pd.DataFrame(self._dataset[:SAMPLED_STRING_ROWS, :], dtype=self._dataset.dtype)
+        else:
+            tmp_dataset = tmp_dataset.iloc[:SAMPLED_STRING_ROWS]
         categorical_col_names = list(np.array(list(tmp_dataset))[(tmp_dataset.applymap(type) == str).all(0)])
         if categorical_col_names:
             all_columns = tmp_dataset.columns
```
Comment (gaugup) [inline python/interpret_community/dataset/dataset_wrapper.py:328]: Might be better to move all the logic for categorical case into a function.

File: python/interpret_community/mimic/mimic_explainer.py [hunk 2]
```diff
@@ -239,9 +239,12 @@ def __init__(self, model, initialization_examples, explainable_model, explainabl
...
         self._datamapper = None
         if transformations is not None:
-            self._datamapper, initialization_examples = get_datamapper_and_transformed_data(
+            self._datamapper, new_initialization_examples = get_datamapper_and_transformed_data(
                 examples=initialization_examples, transformations=transformations,
                 allow_all_transformations=allow_all_transformations)
+            # memory optimization to clear all internal references so they can be gc'ed as soon as possible
+            initialization_examples._clear()
+            initialization_examples = new_initialization_examples
         if reset_index != ResetIndex.Ignore:
```
Comment (gaugup) [inline python/interpret_community/mimic/mimic_explainer.py:248]: Shouldn't python garbage collect if initialization_examples is not longer in use. Why do we need to clear this here?

**Review by gaugup — APPROVED** [review]

File: python/interpret_community/mimic/mimic_explainer.py (hunk 2 above)
Comment (imatiach-msft) [reply python/interpret_community/mimic/mimic_explainer.py:248]: yes, it is garbage collected after designer wrapper loses reference to passed initialization_examples, but we can free the memory much, much earlier by clearing the references after we are done here, thereby reducing overall memory usage in that scenario by 2 GB

File: python/interpret_community/dataset/dataset_wrapper.py (hunk 1 above)
Comment (imatiach-msft) [reply python/interpret_community/dataset/dataset_wrapper.py:328]: this string_index function contains categorical logic only anyway

**Review by gregorybchris — APPROVED** [review]

**Pull request merged**